### PR TITLE
Fixed #31351 -- Added system checks for partial indexes and unique constraints support.

### PR DESCRIPTION
**Pull request by ichlaffterlalu**
Ticket link: https://code.djangoproject.com/ticket/31351
This pull request added warning to alert users that use MySQL/MariaDB or other DBs that doesn't support partial indexing, when they are using UniqueConstrant with conditions. The behavior is actually the same: Django will ignore the constraint, but this time it wouldn't be doing it silently.

### CODE REVIEW

File: django/db/models/base.py [hunk 1]
```diff
@@ -1861,6 +1861,33 @@ def _check_constraints(cls, databases):
...
+            ):
+                continue
+            if any(isinstance(constraint, UniqueConstraint) for constraint in cls._meta.constraints):
+                errors.append(
+                    checks.Warning(
+                        "%s does not support partial indexing, so the unique constraint "
+                        "won't be created." % connection.display_name,
+                        hint=(
+                            "The constraint won't be created. Silence this "
+                            "warning if you don't care about it."
```
Comment (matthijskooijman) [inline django/db/models/base.py]: The wording is a bit inconsistent with the one for check constraints, here there is some duplicate info between the warning and the hint, and it talks about "The constraint" without naming it. I would suggest:

```
                    checks.Warning(
                        "%s does not support unique constraints with conditions." % connection.display_name,
                        hint=(
                            "A constraint won't be created. Silence this "
                            "warning if you don't care about it."
```

File: docs/ref/checks.txt [hunk 2]
```diff
@@ -330,6 +330,8 @@ Models
   ``<max_length>`` characters.
 * **models.W035**: ``db_table`` ``<db_table>`` is used by multiple models:
   ``<model list>``.
+* **models.W036**: ``<database>`` does not support partial indexing, so the
+  unique constraint won't be created.
```
Comment (matthijskooijman) [inline docs/ref/checks.txt]: If the message shown is updated as I suggested, this should probably follow that wording.

File: tests/invalid_models_tests/test_models.py [hunk 3]
```diff
@@ -1325,3 +1325,38 @@ class Meta:
...
+            "won't be created." % connection.display_name,
+            hint=(
+                "A constraint won't be created. Silence this warning if you "
+                "don't care about it."
+            ),
+            obj=Model,
+            id='models.W027',
+        )
+        expected = [] if connection.features.supports_partial_indexes else [warn]
+        self.assertCountEqual(errors, expected)
```
Comment (matthijskooijman) [inline tests/invalid_models_tests/test_models.py]: This should use `assertEqual` and not `assertCountEqual`, otherwise there is no point in entirely recreating the expected warning, you could just hardcode 1 or 0.

File: tests/invalid_models_tests/test_models.py [hunk 4]
```diff
@@ -1325,3 +1325,38 @@ class Meta:
...
+        errors = Model.check(databases=self.databases)
+        warn = Warning(
+            "%s does not support partial indexing, so the unique constraint "
+            "won't be created." % connection.display_name,
+            hint=(
+                "A constraint won't be created. Silence this warning if you "
+                "don't care about it."
+            ),
+            obj=Model,
+            id='models.W027',
```
Comment (matthijskooijman) [inline tests/invalid_models_tests/test_models.py]: This warning ID was not updated after copy-pasting it.

File: django/db/models/base.py [hunk 5]
```diff
@@ -1861,6 +1861,33 @@ def _check_constraints(cls, databases):
                 )
         return errors
 
+    @classmethod
+    def _warn_unique_constraints(cls, databases):
+        errors = []
+        for db in databases:
+            if not router.allow_migrate_model(db, cls):
+                continue
+            connection = connections[db]
```
Comment (matthijskooijman) [inline django/db/models/base.py]: I would suggest integrating this code into `_check_constraints` above, since then the duplicate code up to here can be shared. The name of that function implies that it is expected to check *all* constraints (I' reading it as "check all constraints", not "do something about check-constraints"). There might even be some code sharing in the error generating (e.g. if the constraint-specific check code sets a `message` and `code` variable, some common code could fill in the rest of the warning (especially if the hints are made equal as I suggest below).

Comment (matthijskooijman) [reply django/db/models/base.py]: Also, it seems that you are not actually calling your new function yet, which should be fixed (or will be automatically fixed if you integrate into `_check_constraints`, which I would prefer).

**Comment by matthijskooijman** [comment]
I just noticed that the django [`Index` class also supports a condition](https://docs.djangoproject.com/en/3.0/ref/models/indexes/#condition) that are only supported when `supports_partial_indexes` is set. I think this one would actually warrant a similar warning as well. I don't think all three of these warnings should end up in the same check function (since this would check an index, not a constraint),  but I'm not sure whether it would be better to group the constraint checks together, or group the two partial index checks together...

File: django/db/models/base.py (hunk 5 above)
Comment (ichlaffterlalu) [reply django/db/models/base.py]: After applying your suggestion, it fails at MariaDB. Seems like MariaDB supports check constraint but doesn't support partial indexing. So, I conclude that the `_check_constraints` method is about "do something about check-constraints", and in the latest commit (testing job ongoing right now) I resplitted back to `_check_unique_constraints` method.

Comment (matthijskooijman) [reply django/db/models/base.py]: My suggestion was to integrate both in the same method, but in that same method look at `supports_table_check_constraints` for `CheckConstraints` as before, and look at `support_partial_indexes` for `UniqueConstraints` with `condition`. I was not trying to suggest that both should only look at `supports_table_check_constraints`, but maybe you did think I did?

**Comment by matthijskooijman** [comment]
The patch as-is looks good to me. However, I do still see two possible improvements, as suggested before.
 - Remove some duplicate code between the `_check_constraints` and `_check_unique_constraints` functions by merging them. You previously tried, but that did not work because you did not rewrite the `continue` statement in there (should be changed to an if around the check rather than an if with continue before the check).
 - Also check for `Index` with conditions and produce a warning when they are used when `supports_partial_indexes` is `False`. This probably is best in a separate function, since there will be only just a little bit of common code.

It's not up to me to decide whether these improvements are needed for merging, though.

**Comment by ichlaffterlalu** [comment]
Umm, is it okay to change the CheckConstraint related checks, when this ticket actually has nothing to do with that code? I don't know whether it should go for another ticket or not at Django.

For Index with conditions, I'll try to figure out on how to also check them.

**Comment by matthijskooijman** [comment]
> Umm, is it okay to change the CheckConstraint related checks, when this ticket actually has nothing to do with that code? I don't know whether it should go for another ticket or not at Django.

If that would prevent duplicate code, then it is perfectly acceptable to change code as part of adding something new. You're not changing the *behaviour* of the existing check, just reusing some of the code. That's just good software engineering practice.

**Comment by matthijskooijman** [comment]
Last push looks good. Did not realize there was already a `_check_indexes`, looks like the right place for this addition.

One thing though, IIUC the index check now happens against the default connection only, regardless of migration routing. E.g. this bit of code should probably be in `_check_indexes` too (only around the new check, not the existing checks):

https://github.com/django/django/blob/627df13faad49a5feb8ccd3e55eb633fa23bf8a4/django/db/models/base.py#L1857-L1860

File: django/db/models/base.py [hunk 6]
```diff
@@ -1277,7 +1277,7 @@ def check(cls, **kwargs):
                 *cls._check_unique_together(),
                 *cls._check_indexes(),
```
Comment (charettes) [inline django/db/models/base.py]: This needs to be based of `databases` just like `_check_constraints` is.

File: django/db/models/base.py [hunk 7]
```diff
@@ -1608,6 +1608,22 @@ def _check_indexes(cls):
                         id='models.E034',
                     ),
                 )
+            if not (
+                connection.features.supports_partial_indexes or
```
Comment (charettes) [inline django/db/models/base.py]: You can't use `connection` here, you need to pass `databases` to `_check_indexes` and follow what `_check_constraints` does.

Comment (ichlaffterlalu) [reply django/db/models/base.py]: Hmm, seems like I need to separate the check into another function.

**Comment by ichlaffterlalu** [comment]
Seems like there's inconsistencies on what database settings list to use. In `_check_long_column_names`, database list are retrieved from `django.conf.settings.DATABASES`. In `_check_constraints`, it uses `database` from `kwargs` at `check(cls, **kwargs)` method. The weird thing is, when I followed the same way others did at `_check_constraints`, it failed at existing test. So, I prefer to use `settings.DATABASES` as the source of database list.

**Comment by matthijskooijman** [comment]
> Seems like there's inconsistencies on what database settings list to use. In _check_long_column_names, database list are retrieved from django.conf.settings.DATABASES. In _check_constraints, it uses database from kwargs at check(cls, **kwargs) method.
The weird thing is, when I followed the same way others did at _check_constraints, it failed at existing test. So, I prefer to use settings.DATABASES as the source of database list.

Hm, I'm not sure if this is the correct approach. `settings.DATABASES` will always contain *all* configured databases, but the check command might be run for a subset. This code was fairly recently introduced (See #12396 and #12432), so maybe its implementation missed a few cases.

I would suggest going back to using the passed-in list of databases, and then fixing the test failure that results from that. Do you remember which test failed?

As for the inconsistency with `_check_long_column_names`, I actually think the latter is actually also incorrect, which I pointed out in a separate comment at https://github.com/django/django/pull/12396#issuecomment-600550555

> UPDATE: somehow tests are successfully finished on my PC but failed on CI. I think it happens because on the MySQL/MariaDB Django CI, they use 2 database connections on test settings: one with MySQL, and one with MariaDB, so the warning also shown twice. I will update my tests to use assertIn instead.

I think this analysis is not entirely correct: I suspect *all* tests use two database connections, one for the `default` database and one for the `other` database. However, this warning is only generated for MySQL/MariaDB, so that explains why none of the other tests see this. In this case, I also think the warning might also be caused by the change from the passed in databases (which *might* default to only `default`, I think) to `settings.DATABASES` (which will always contain both). Nevertheless, having `assertIn` seems ok to me, since there might be other warnings as well in the future.

**Comment by ichlaffterlalu** [comment]
I'm sorry that I didn't respond to suggestions for so long. I finally found the issue is actually `IndexesTests` test suite inherits `SimpleTestCase` instead of `TestCase` like `ConstraintsTests` did. That's why the `database` parameter wasn't initialized at all and it fails. Now that I reverted my changes back to using `database` parameter instead of `settings.DATABASE`, I think my current work is complete, unless if there's anything I still missed of.

P.S. I have no idea regarding `docs` check failure. I didn't change any of the documentations in recent commits. Last commit was green on that check.

**Comment by charettes** [comment]
> P.S. I have no idea regarding docs check failure. I didn't change any of the documentations in recent commits. Last commit was green on that check.

It's not due to your changes, it should be addressed by https://github.com/django/django/pull/12670

File: tests/invalid_models_tests/test_models.py [hunk 8]
```diff
@@ -1313,8 +1356,8 @@ class Meta:
             obj=Model,
             id='models.W027',
         )
-        expected = [] if connection.features.supports_table_check_constraints else [warn]
-        self.assertCountEqual(errors, expected)
+        if not connection.features.supports_table_check_constraints:
+            self.assertIn(warn, errors)
```
Comment (felixxm) [inline tests/invalid_models_tests/test_models.py:1360]: This change is not related with this patch, please revert it.

File: django/db/models/base.py [hunk 9]
```diff
@@ -1608,6 +1608,28 @@ def _check_indexes(cls):
                         id='models.E034',
                     ),
                 )
+
+        for db in databases:
+            print("hehe")
```
Comment (felixxm) [inline django/db/models/base.py:1613]: Please remove prints.

File: django/db/models/base.py [hunk 10]
```diff
@@ -1275,9 +1275,9 @@ def check(cls, **kwargs):
             errors += [
                 *cls._check_index_together(),
                 *cls._check_unique_together(),
-                *cls._check_indexes(),
                 *cls._check_ordering(),
-                *cls._check_constraints(databases),
+                *cls._check_indexes(databases),
+                *cls._check_constraints(databases)
```
Comment (felixxm) [inline django/db/models/base.py:1280]: Add trailing comma.

File: django/db/models/base.py [hunk 11]
```diff
@@ -1275,9 +1275,9 @@ def check(cls, **kwargs):
             errors += [
                 *cls._check_index_together(),
                 *cls._check_unique_together(),
-                *cls._check_indexes(),
                 *cls._check_ordering(),
-                *cls._check_constraints(databases),
+                *cls._check_indexes(databases),
```
Comment (felixxm) [inline django/db/models/base.py:1279]: Is there any reason to move `_check_indexes()`?

File: tests/invalid_models_tests/test_models.py (hunk 8 above)
Comment (matthijskooijman) [reply tests/invalid_models_tests/test_models.py:1360]: Hm, I think I suggested this change on the new test, did not realize the existing test used `assertCountEqual`. Still, this looks like a good improvement, but then in a separate commit or PR (the commit history of this PR should probably be squashed anyway before merging).

Comment (felixxm) [reply tests/invalid_models_tests/test_models.py:1360]: It's not an improvement. We run tests on databases that support check constraints and on databases that don't support them. We want to check both cases.

Comment (matthijskooijman) [reply tests/invalid_models_tests/test_models.py:1360]: Right, I think the improvement part is that you're actually looking at the content of the errors, rather than just at the count. With the original code, there really is no point in recreating the full warning literally, you could also just say:

```
expected = 0 if connection.features.supports_table_check_constraints else 1
self.assertEqual(len(errors), expected)
```

Actually checking the content of the error sounds like an improvement to me. But the proposed change actually checks *nothing* when the database does not support check constraints, which is indeed not good.

So, how about just replacing the `assertCountEqual` with an `assertEqual`, e.g.:

```
expected = [] if connection.features.supports_table_check_constraints else [warn]
        self.assertEqual(errors, expected)
```

Using `assertIn` rather than `assertEqual` also seemed reasonable (since there might be more warnings (maybe not now, then later). But then you would need to check either `assertIn` or `assertNotIn`, e.g.:

```
if not connection.features.supports_table_check_constraints:
            self.assertIn(warn, errors)
else:
            self.assertNotIn(warn, errors)
```

This seems to be the most complete and generic check?

Comment (felixxm) [reply tests/invalid_models_tests/test_models.py:1360]: > Right, I think the improvement part is that you're actually looking at the content of the errors, rather than just at the count. With the original code, there really is no point in recreating the full warning literally, you could also just say:
> 
> ```
> expected = 0 if connection.features.supports_table_check_constraints else 1
> self.assertEqual(len(errors), expected)
> ```

[`assertCountEqual()`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual) doesn't work that way, but do to the fact that we have list with only 1 element we can use `self.assertEqual(errors, expected)`.

File: django/db/models/base.py (hunk 11 above)
Comment (ichlaffterlalu) [reply django/db/models/base.py:1279]: There is no particular reason for this change. At previous commits I try to address errors that showed up in GitHub check but not in my computer. I'll revert that.

File: tests/invalid_models_tests/test_models.py (hunk 8 above)
Comment (ichlaffterlalu) [reply tests/invalid_models_tests/test_models.py:1360]: So, is `assertCountEqual()` like counting how many elements inside a list that equals to the object we're checking, not the whole length of list?
If that's so, I think using `assertCountEqual()` is a good option.

Comment (matthijskooijman) [reply tests/invalid_models_tests/test_models.py:1360]: Oh, confusing name then, `assertCountEqual()`, thanks for the link.

So then this is really not much of an improvement, unless the original code fails because there is more than one one error (but given only a very specific model is checked, it should not normally raise additional warnings, and if it does, failing the test to force review of the testcase is probably a good idea).

So indeed, reverting this change would be best. Sorry for the noise :-)

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@ichlaffterlalu Thanks :+1: I pushed minor edits.

**Comment by felixxm** [comment]
@ichlaffterlalu Welcome aboard :boat:

**Comment by matthijskooijman** [comment]
@ichlaffterlalu, thanks for picking up my issue and persisting through all our comments :-D
